File: backend/python-api/app/service/auth_service.py
from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass
from datetime import datetime, timezone

from passlib.hash import bcrypt

from app.domain.errors import (
    ERR_CONFLICT,
    ERR_INVALID_INPUT,
    ERR_UNAUTHORIZED,
    DomainError,
    TwoFARequiredError,
)
from app.domain.password import validate_password_complexity
from app.domain.token import EmailVerificationToken, generate_secure_token, token_expiry
from app.domain.username import normalize_username, validate_username
from app.domain.user import User
from app.dto.auth import LoginRequest, RegisterRequest, TokenPair
from app.i18n.loader import is_language_supported
from app.platform.captcha import CaptchaVerifier
from app.platform.email import EmailSender
from app.platform.geoip import GeoIPResolver
from app.repository.login_event_repo import LoginEventRepository
from app.repository.settings_repo import SettingsRepository
from app.repository.token_repo import TokenRepository
from app.repository.user_repo import UserRepository
from app.service.token_service import TokenService
from app.domain.login_event import LoginEvent

logger = logging.getLogger(__name__)


@dataclass
class RegisterResult:
    user: User
    verification_required: bool


@dataclass
class LoginResult:
    user: User
    token_pair: TokenPair


class AuthService:
    def __init__(
        self,
        users: UserRepository,
        tokens: TokenRepository,
        login_events: LoginEventRepository,
        settings: SettingsRepository,
        token_svc: TokenService,
        email_sender: EmailSender,
        captcha: CaptchaVerifier,
        geoip: GeoIPResolver,
        frontend_url: str,
    ):
        self._users = users
        self._tokens = tokens
        self._login_events = login_events
        self._settings = settings
        self._token_svc = token_svc
        self._email = email_sender
        self._captcha = captcha
        self._geoip = geoip
        self._frontend_url = frontend_url

    async def register(self, req: RegisterRequest, client_ip: str, session) -> RegisterResult:
        if req.recaptcha_token:
            ok = await self._captcha.verify(req.recaptcha_token)
            if not ok:
                raise DomainError(ERR_INVALID_INPUT, "reCAPTCHA verification failed")

        username = normalize_username(req.username) if req.username else ""
        if username:
            err = validate_username(username)
            if err:
                raise DomainError(ERR_INVALID_INPUT, err)
            available = await self._users.is_username_available(username)
            if not available:
                raise DomainError(ERR_CONFLICT, "Username already taken")

        existing = await self._users.find_by_email_unscoped(req.email)
        if existing and not existing.is_deleted:
            raise DomainError(ERR_CONFLICT, "User already exists")

        err = validate_password_complexity(req.password)
        if err:
            raise DomainError(ERR_INVALID_INPUT, err)

        lang = self._normalize_lang(req.language)
        country = await self._resolve_country(req.country, client_ip, lang)
        password_hash = bcrypt.using(rounds=12).hash(req.password)

        require_verification = await self._is_verification_required()

        user = User(
            username=username or None,
            first_name=req.first_name,
            last_name=req.last_name,
            email=req.email,
            password_hash=password_hash,
            role="user",
            language=lang,
            country=country,
            verified=not require_verification,
        )

        if existing and existing.is_deleted:
            await self._users.hard_delete_related_records(existing.id)
            await self._users.hard_delete(existing.id)

        await self._users.create(user)

        verification_token_str = ""
        if require_verification:
            verification_token_str = generate_secure_token()
            v_token = EmailVerificationToken(
                user_id=user.id,
                token=verification_token_str,
                expires_at=token_expiry(15),
            )
            await self._tokens.create_email_verification_token(v_token)

        await session.commit()

        if require_verification and verification_token_str:
            link = f"{self._frontend_url}/verify-email?token={verification_token_str}"
            asyncio.create_task(self._email.send_verification(user.email, link, lang, user.full_name))

        return RegisterResult(user=user, verification_required=require_verification)

    async def login(self, req: LoginRequest, client_ip: str, user_agent: str, session) -> LoginResult:
        if req.recaptcha_token:
            ok = await self._captcha.verify(req.recaptcha_token)
            if not ok:
                raise DomainError(ERR_INVALID_INPUT, "reCAPTCHA verification failed")

        try:
            user = await self._users.find_by_email_or_username(req.email_or_username)
        except DomainError:
            raise DomainError(ERR_UNAUTHORIZED, "Invalid credentials")

        if not bcrypt.verify(req.password, user.password_hash):
            raise DomainError(ERR_UNAUTHORIZED, "Invalid credentials")

        if not user.verified:
            raise DomainError(ERR_UNAUTHORIZED, "Please verify your email address before logging in")

        if user.two_fa_enabled:
            if user.two_fa_secret:
                raise TwoFARequiredError("totp", user.public_id, "TOTP code required")
            else:
                from app.service.twofa_service import TwoFAService
                raise TwoFARequiredError("email", user.public_id, "2FA code sent to email")

        pair = await self._token_svc.issue_token_pair(user.id, user.role, req.remember_me)

        event = LoginEvent(user_id=user.id, ip=client_ip, user_agent=user_agent)
        await self._login_events.create(event)
        await session.commit()

        return LoginResult(user=user, token_pair=pair)

    async def get_me(self, user_id: int) -> User:
        return await self._users.find_by_id(user_id)

    async def refresh_token(self, raw_refresh_token: str, session) -> TokenPair:
        user_id, family_id = await self._token_svc.validate_and_revoke_refresh_token(raw_refresh_token)
        user = await self._users.find_by_id(user_id)
        pair = await self._token_svc.issue_rotated_pair(user.id, user.role, family_id)
        await session.commit()
        return pair

    async def logout(self, raw_refresh_token: str, session) -> None:
        await self._token_svc.revoke_refresh_token(raw_refresh_token)
        await session.commit()

    async def logout_all(self, user_id: int, session) -> None:
        await self._token_svc.revoke_all_user_tokens(user_id)
        await session.commit()

    async def _is_verification_required(self) -> bool:
        val = await self._settings.get("email_verification_required")
        return val.lower() == "true" if val else False

    def _normalize_lang(self, lang: str) -> str:
        if lang and is_language_supported(lang.lower()):
            return lang.lower()
        return "en"

    async def _resolve_country(self, country: str, client_ip: str, lang: str) -> str:
        if country:
            return country.upper()[:10]
        resolved = await self._geoip.resolve_country(client_ip)
        return resolved or ""

File: backend/go-api/internal/service/auth_service.go
package service

import (
	"context"
	"errors"
	"log"
	"strings"
	"time"

	"saas-starter/backend/go-api/internal/domain"
	"saas-starter/backend/go-api/internal/dto"
	"saas-starter/backend/go-api/internal/platform/captcha"
	"saas-starter/backend/go-api/internal/platform/email"
	"saas-starter/backend/go-api/internal/platform/geoip"
	"saas-starter/backend/go-api/internal/repository"
	"saas-starter/backend/go-api/pkg/i18n"

	"golang.org/x/crypto/bcrypt"
)

type RegisterResult struct {
	User                 *domain.User
	VerificationRequired bool
}

type LoginResult struct {
	User      *domain.User
	TokenPair *dto.TokenPair
}

type AuthService struct {
	users       repository.UserRepository
	tokens      repository.TokenRepository
	loginEvents repository.LoginEventRepository
	settings    repository.SettingsRepository
	txManager   repository.TxManager
	tokenSvc    *TokenService
	emailSender email.Sender
	captcha     captcha.Verifier
	geoIP       geoip.Resolver
	frontendURL string
}

func NewAuthService(
	users repository.UserRepository,
	tokens repository.TokenRepository,
	loginEvents repository.LoginEventRepository,
	settings repository.SettingsRepository,
	txManager repository.TxManager,
	tokenSvc *TokenService,
	emailSender email.Sender,
	captchaVerifier captcha.Verifier,
	geoIPResolver geoip.Resolver,
	frontendURL string,
) *AuthService {
	return &AuthService{
		users:       users,
		tokens:      tokens,
		loginEvents: loginEvents,
		settings:    settings,
		txManager:   txManager,
		tokenSvc:    tokenSvc,
		emailSender: emailSender,
		captcha:     captchaVerifier,
		geoIP:       geoIPResolver,
		frontendURL: frontendURL,
	}
}

func (s *AuthService) Register(ctx context.Context, req dto.RegisterRequest, clientIP string) (*RegisterResult, error) {
	if req.RecaptchaToken != "" {
		if err := s.captcha.Verify(req.RecaptchaToken, "register"); err != nil {
			return nil, domain.NewError(domain.ErrInvalidInput, "reCAPTCHA verification failed")
		}
	}

	username := domain.NormalizeUsername(req.Username)
	if err := domain.ValidateUsername(username); err != nil {
		return nil, domain.NewError(domain.ErrInvalidInput, err.Error())
	}

	available, err := s.users.IsUsernameAvailable(ctx, username)
	if err != nil {
		return nil, err
	}
	if !available {
		return nil, domain.NewError(domain.ErrConflict, "username already taken")
	}

	existing, err := s.users.FindByEmailUnscoped(ctx, req.Email)
	if err != nil && !errors.Is(err, domain.ErrNotFound) {
		return nil, err
	}
	if existing != nil && !existing.DeletedAt.Valid {
		return nil, domain.NewError(domain.ErrConflict, "user already exists")
	}

	if err := domain.ValidatePasswordComplexity(req.Password); err != nil {
		return nil, domain.NewError(domain.ErrInvalidInput, err.Error())
	}

	lang := normalizeLang(req.Language)
	country := s.resolveCountry(req.Country, clientIP, lang)

	hash, err := bcrypt.GenerateFromPassword([]byte(req.Password), BcryptCost)
	if err != nil {
		return nil, err
	}

	requireVerification := s.isVerificationRequired(ctx)
	require2FA := s.is2FARequired(ctx)

	user := &domain.User{
		Username:     username,
		FirstName:    req.FirstName,
		LastName:     req.LastName,
		Email:        req.Email,
		PasswordHash: string(hash),
		Role:         "user",
		Language:     lang,
		Country:      country,
		Verified:     !requireVerification,
		TwoFAEnabled: require2FA,
	}

	var verificationTokenStr string

	err = s.txManager.WithTx(ctx, func(txCtx context.Context) error {
		if existing != nil && existing.DeletedAt.Valid {
			if err := s.users.HardDeleteRelatedRecords(txCtx, existing.ID); err != nil {
				return err
			}
			if err := s.users.HardDelete(txCtx, existing.ID); err != nil {
				return err
			}
		}

		if err := s.users.Create(txCtx, user); err != nil {
			return err
		}

		if requireVerification {
			verificationTokenStr = domain.GenerateSecureToken()
			vToken := &domain.EmailVerificationToken{
				UserID:    user.ID,
				Token:     verificationTokenStr,
				ExpiresAt: domain.TokenExpiry(15),
			}
			return s.tokens.CreateEmailVerificationToken(txCtx, vToken)
		}
		return nil
	})

	if err != nil {
		return nil, err
	}

	if requireVerification && verificationTokenStr != "" {
		go func() {
			link := s.frontendURL + "/verify-email?token=" + verificationTokenStr
			if sendErr := s.emailSender.SendVerification(user.Email, link, lang, user.FullName()); sendErr != nil {
				log.Printf("Failed to send verification email to %s: %v", user.Email, sendErr)
			}
		}()
	}

	return &RegisterResult{User: user, VerificationRequired: requireVerification}, nil
}

func (s *AuthService) Login(ctx context.Context, req dto.LoginRequest, clientIP, userAgent string) (*LoginResult, error) {
	if req.RecaptchaToken != "" {
		if err := s.captcha.Verify(req.RecaptchaToken, "login"); err != nil {
			return nil, domain.NewError(domain.ErrInvalidInput, "reCAPTCHA verification failed")
		}
	}

	input := strings.TrimSpace(req.EmailOrUsername)
	normalized := strings.ToLower(input)

	var user *domain.User
	var err error
	if strings.Contains(normalized, "@") {
		user, err = s.users.FindByEmail(ctx, normalized)
	} else {
		user, err = s.users.FindByUsername(ctx, normalized)
	}
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NewError(domain.ErrUnauthorized, "invalid credentials")
		}
		return nil, err
	}

	if bcryptErr := bcrypt.CompareHashAndPassword([]byte(user.PasswordHash), []byte(req.Password)); bcryptErr != nil {
		return nil, domain.NewError(domain.ErrUnauthorized, "invalid credentials")
	}

	if s.isVerificationRequired(ctx) && !user.Verified {
		return nil, domain.NewError(domain.ErrEmailNotVerified, "email not verified")
	}

	if user.TwoFAEnabled {
		twoFAType := "email"
		if user.TwoFASecret != "" {
			twoFAType = "totp"
		}

		if twoFAType == "email" {
			s.send2FACode(ctx, user, req.RememberMe)
		}

		return nil, &domain.TwoFARequiredError{
			PublicID:  user.PublicID,
			TwoFAType: twoFAType,
			Message:   "2FA required",
		}
	}

	pair, err := s.tokenSvc.IssueTokenPair(ctx, user.ID, user.Role, req.RememberMe)
	if err != nil {
		return nil, err
	}

	go s.recordLoginEvent(ctx, user.ID, clientIP, userAgent)

	return &LoginResult{User: user, TokenPair: pair}, nil
}

func (s *AuthService) GetMe(ctx context.Context, userID uint) (*domain.User, error) {
	return s.users.FindByID(ctx, userID)
}

func (s *AuthService) RefreshToken(ctx context.Context, rawRefreshToken string) (*dto.RefreshTokenResponse, error) {
	result, err := s.tokenSvc.ValidateAndRevokeRefreshToken(ctx, rawRefreshToken)
	if err != nil {
		return nil, err
	}

	user, err := s.users.FindByID(ctx, result.UserID)
	if err != nil {
		return nil, err
	}

	pair, err := s.tokenSvc.IssueRotatedPair(ctx, user.ID, user.Role, result.FamilyID)
	if err != nil {
		return nil, err
	}

	return &dto.RefreshTokenResponse{
		AccessToken:  pair.AccessToken,
		RefreshToken: pair.RefreshToken,
		ExpiresIn:    pair.ExpiresIn,
		TokenType:    pair.TokenType,
	}, nil
}

func (s *AuthService) Logout(ctx context.Context, rawRefreshToken string) error {
	return s.tokenSvc.RevokeRefreshToken(ctx, rawRefreshToken)
}

func (s *AuthService) LogoutAllSessions(ctx context.Context, userID uint) error {
	return s.tokenSvc.RevokeAllUserTokens(ctx, userID)
}

// send2FACode is a placeholder — the actual 2FA code sending
// is handled by TwoFAService. The auth service only detects that
// 2FA is required and returns TwoFARequiredError to the handler,
// which then delegates to the twofa handler/service flow.
func (s *AuthService) send2FACode(_ context.Context, _ *domain.User, _ bool) {
	// Intentionally empty: 2FA email sending is handled by TwoFAService.Resend2FA
	// which the frontend calls after receiving requires_2fa=true.
}

func (s *AuthService) recordLoginEvent(ctx context.Context, userID uint, ip, userAgent string) {
	event := &domain.LoginEvent{
		UserID:    userID,
		IP:        ip,
		UserAgent: userAgent,
		LoggedAt:  time.Now(),
	}
	if err := s.loginEvents.Create(context.Background(), event); err != nil {
		log.Printf("Failed to record login event for user %d: %v", userID, err)
	}
}

func (s *AuthService) isVerificationRequired(ctx context.Context) bool {
	val := s.settings.Get(ctx, "require_email_verification", "true")
	return val == "true"
}

func (s *AuthService) is2FARequired(ctx context.Context) bool {
	val := s.settings.Get(ctx, "require_2fa", "false")
	return val == "true"
}

func (s *AuthService) resolveCountry(country, clientIP, lang string) string {
	if country != "" {
		c := strings.ToUpper(country)
		if len(c) > 10 {
			c = c[:10]
		}
		return c
	}
	if detected := s.geoIP.CountryFromIP(clientIP); detected != "" {
		return detected
	}
	return s.geoIP.CountryFromLanguage(lang)
}

func normalizeLang(lang string) string {
	if lang == "" {
		return "en"
	}
	lang = strings.ToLower(lang)
	if len(lang) > 2 {
		lang = lang[:2]
	}
	if !i18n.IsLanguageSupported(lang) {
		return "en"
	}
	return lang
}

